src/sentry/sentry_metrics/querying/data/transformation.py

```diff
@@ -0,0 +1,218 @@
+from sentry.sentry_metrics.querying.data.utils import get_identity, nan_to_none
+from sentry.sentry_metrics.querying.errors import (
+    MetricsQueryExecutionError,
+)
+from sentry.sentry_metrics.querying.types import (
+    GroupKey,
+    ResultValue,
+    Series,
+    Total,
+)
+
+
+@dataclass
+class GroupValue:
+    series: Series
+    total: Total
+
+    @classmethod
+    def empty(cls) -> "GroupValue":
+        return GroupValue(series=[], total=None)
+
+    def add_series_entry(self, time: str, aggregate_value: ResultValue):
+        self.series.append((time, self._transform_aggregate_value(aggregate_value)))
+
+    def add_total(self, aggregate_value: ResultValue):
+        self.total = self._transform_aggregate_value(aggregate_value)
+
+    def _transform_aggregate_value(self, aggregate_value: ResultValue):
+        # For now, we don't support the array return type, since the set of operations that the API can support
+        # won't lead to multiple values in a single aggregate value. For this reason, we extract the first value
+        # in case we get back an array of values, which can happen for multiple quantiles.
+        if isinstance(aggregate_value, list):
+            if aggregate_value:
+                return aggregate_value[0]
+
+            raise MetricsQueryExecutionError("Received an empty array as aggregate value")
+
+        return aggregate_value
+
+
+@dataclass
+class QueryMeta:
+    name: str
+    type: str
+
+    def __post_init__(self):
+        self._transform_meta_type()
+
+    def _transform_meta_type(self):
+        # Since we don't support the array aggregate value, and we return the first element, we just return the type of
+        # the values of the array.
+        if self.type.startswith("Array("):
+            self.type = self.type[6 : len(self.type) - 1]
+
+
+def _build_intervals(start: datetime, end: datetime, interval: int) -> Sequence[datetime]:
+    """
+    Builds a list of all the intervals that are queried by the metrics layer.
+    """
+    start_seconds = start.timestamp()
+    end_seconds = end.timestamp()
+
+    current_time = start_seconds
+    intervals = []
+    while current_time < end_seconds:
+        intervals.append(datetime.fromtimestamp(current_time, timezone.utc))
+        current_time = current_time + interval
+
+    return intervals
+
+
+def _generate_full_series(
+    start_seconds: int,
+    num_intervals: int,
+    interval: int,
+    series: Series,
+    null_value: ResultValue = None,
+) -> Sequence[ResultValue]:
+    """
+    Computes a full series over the entire requested interval with None set where there are no data points.
+    """
+    full_series = [null_value] * num_intervals
+    for time, value in series:
+        time_seconds = parse_datetime_string(time).timestamp()
+        index = int((time_seconds - start_seconds) / interval)
+        full_series[index] = value
+
+    return full_series
+
+class QueryTransformer:
```

Comment:
This is the only thing that was changed. Instead of a function, I used a class which contains some shared state that looks a bit better, even though I am not very happy with it. I won't do much work here since we will likely have to change how the results are formatted when multiseries support will be added.